### ISSUE TITLE
Remove dependency on libgcc_s_sjlj-1.dll for 32bit versions

### ISSUE THREAD

**Issue by k-takata** [mlocati/gettext-iconv-windows#8]
As we discussed in vim/vim-win32-installer#20 (and #7), `libintl-8.dll` has dependency on `libgcc_s_sjlj-1.dll`.
I thought the dependency should be able to be removed by adding `-static-libgcc`, but it didn't work, as you know.

I installed your build script (with mxe) and after doing some trials, I found the cause of the issue: The `libtool` script stripped the `-static-libgcc` option.

When I edited the Makefile for `libintl-8.dll` as follows, the dependency was successfully removed:

```diff
--- build-gettext-windows/shared-32/configured/gettext-0.19.8.1/gettext-runtime/intl/Makefile.org
+++ build-gettext-windows/shared-32/configured/gettext-0.19.8.1/gettext-runtime/intl/Makefile
@@ -108,7 +108,7 @@
 LDFLAGS = -L/home/ken/build-gettext-windows/shared-32/compiled/gettext/lib -Wl,--disable-auto-import $(LDFLAGS_yes)
 LDFLAGS_yes = -Wl,--export-all-symbols
 LDFLAGS_no =
-LIBS = 
+LIBS = -XCClinker -static-libgcc
 
 COMPILE = $(CC) -c $(DEFS) $(INCLUDES) $(CPPFLAGS) $(CFLAGS) $(XCFLAGS)
 
```
`-XCClinker` is the option for libtool for passing the following flag to the compiler driver.

I don't know how I can apply this change into your build script, but could you consider applying this?

**Comment by mlocati** [comment]
I can for sure update the package, BTW finding a cleaner solution (or identifying the cause of the issue) may be helpful.
What about asking in the gettext mailing list?

**Comment by k-takata** [comment]
I haven't asked (or even read) the gettext mailing list...

I tired some more to merge the above patch into `compile-iconv-gettext-windows.sh`, but there were many problems.
First, setting `$LIB` like `./configure ... LIB='-XCClinker -static-libgcc' ...` didn't work because `-XCClinker` is the option for libtool but not for gcc. Passing the `$LIB` only to `make` didn't also work.
Patching after configuration to `gettext-runtime/intl/Makefile` under the `configured` directory didn't also work, because the makefile actually used was <code>gettext-<b>tools</b>/intl/Makefile</code>. Moreover patching the file before `make` didn't work, because it was overwritten while the making process.

Finally I think that patching `gettext-runtime/intl/Makefile.in` (before configuration) is the best way to work around this:

```diff
diff --git a/patches/gettext-0.19.8.1-configure/0006-remove-gcc_s_dependency.patch b/patches/gettext-0.19.8.1-configure/0006-remove-gcc_s_dependency.patch
new file mode 100644
index 0000000..3c97ff6
--- /dev/null
+++ b/patches/gettext-0.19.8.1-configure/0006-remove-gcc_s_dependency.patch
@@ -0,0 +1,11 @@
+--- a/gettext-runtime/intl/Makefile.in	2016-06-11 16:37:41.000000000 +0900
++++ b/gettext-runtime/intl/Makefile.in	2018-09-24 01:31:29.479649600 +0900
+@@ -108,7 +108,7 @@
+ LDFLAGS = @LDFLAGS@ $(LDFLAGS_@WOE32DLL@)
+ LDFLAGS_yes = -Wl,--export-all-symbols
+ LDFLAGS_no =
+-LIBS = @LIBS@
++LIBS = @LIBS@ -XCClinker -static-libgcc
+ 
+ COMPILE = $(CC) -c $(DEFS) $(INCLUDES) $(CPPFLAGS) $(CFLAGS) $(XCFLAGS)
+ 
```

This actually works by just executing `compile-iconv-gettext-windows.sh`.

**Comment by k-takata** [comment]
Another way is this: https://github.com/k-takata/gettext-iconv-windows/commit/94691e6c88ae9d84d79cdf9b1b1c32d288997f41

```diff
--- a/compile-iconv-gettext-windows.sh
+++ b/compile-iconv-gettext-windows.sh
@@ -322,6 +322,9 @@ cd gettext-$BLDGTXT_V_GETTEXT
 bldgtxtApplyPatches "$BLDGTXT_PATCHES/gettext-$BLDGTXT_V_GETTEXT-configure"
 mkdir $BLDGTXT_CONFIGURED/gettext-$BLDGTXT_V_GETTEXT
 cd $BLDGTXT_CONFIGURED/gettext-$BLDGTXT_V_GETTEXT
+if [ "$BLDGTXT_LINK-$BLDGTXT_BITS" = "shared-32" ]; then
+    sed -i.orig 's/^LIBS = @LIBS@$/LIBS = @LIBS@ -XCClinker -static-libgcc/' $BLDGTXT_SOURCE/gettext-$BLDGTXT_V_GETTEXT/gettext-runtime/intl/Makefile.in
+fi
 echo '### Configuring gettext'
 $BLDGTXT_SOURCE/gettext-$BLDGTXT_V_GETTEXT/configure \
     --host=$MXE_TARGETS \
```

Either should work as expected.

**Comment by mlocati** [comment]
> I haven't asked (or even read) the gettext mailing list...

I just did it: http://lists.gnu.org/archive/html/bug-gettext/2018-09/msg00024.html

**Comment by mlocati** [comment]
See #9 for a fix of this, but I'd wait for a feedback from the mailing list members to see if there's a cleaner solution.

**Comment by k-takata** [comment]
> I just did it: http://lists.gnu.org/archive/html/bug-gettext/2018-09/msg00024.html

No response for two weeks?

**Comment by k-takata** [comment]
I think #9 is the best way, isn't it?

**Comment by mlocati** [comment]
Yep, I think so.
BTW I'm planning to merge it once we'll have a new gettext version (which [seems to be quite soon](http://lists.gnu.org/archive/html/bug-gettext/2018-10/msg00040.html)).

**Comment by k-takata** [comment]
That's good news.

**Comment by mlocati** [comment]
I tried this patch, but the built `libintl-7.dll` still depends on `libgcc_s_sjlj-1.dll`:

```patch
From: "K.Takata" <kentkt@csc.jp>
Date: Fri, 10 May 2019 14:50:24 +0200
Subject: [PATCH] Remove dependency on libgcc from libintl

--- a/gettext-runtime/intl/Makefile.in
+++ b/gettext-runtime/intl/Makefile.in
@@ -965,7 +965,7 @@ LIBMULTITHREAD = @LIBMULTITHREAD@
 LIBOBJS = @LIBOBJS@
 LIBPTH = @LIBPTH@
 LIBPTH_PREFIX = @LIBPTH_PREFIX@
-LIBS = @LIBS@
+LIBS = @LIBS@ -XCClinker -static-libgcc
 LIBSCHED = @LIBSCHED@
 LIBTHREAD = @LIBTHREAD@
 LIBTOOL = @LIBTOOL@
```

**Comment by k-takata** [comment]
How about this? I don't know the difference between `gettext-runtime` and `gettext-tools`, but it seems that this works.

```diff
--- a/gettext-tools/intl/Makefile.in
+++ b/gettext-tools/intl/Makefile.in
@@ -1473,7 +1473,7 @@
 LIBOBJS = @LIBOBJS@
 LIBPTH = @LIBPTH@
 LIBPTH_PREFIX = @LIBPTH_PREFIX@
-LIBS = @LIBS@
+LIBS = @LIBS@ -XCClinker -static-libgcc
 LIBSOCKET = @LIBSOCKET@
 LIBTESTS_LIBDEPS = @LIBTESTS_LIBDEPS@
 LIBTHREAD = @LIBTHREAD@
```

**Comment by k-takata** [comment]
About the name of `libintl-7.dll`, I think this is a serious bug. If gettext 0.20's API is compatible with 0.19, the file name should keep `libintl-8.dll`. Otherwise it should be `libintl-9.dll`, not `libintl-7.dll`.

This is caused by the `-version-info` option in LDFLAGS set in `gettext-tools/intl/Makefile`.
In 0.19.8.1, it was `-version-info 9:5:1`, but in 0.20, it is `-version-info 9:6:2`.
It should be `-version-info 10:6:2` or something.

**Comment by mlocati** [comment]
I'll check your patch file asap.
About the library version: this isn't really the place to discuss it, you should use the gettext mailing list...

**Comment by k-takata** [comment]
> About the library version: this isn't really the place to discuss it, you should use the gettext mailing list...

OK, I posted a report here: https://savannah.gnu.org/bugs/index.php?56305

**Comment by k-takata** [comment]
gettext 0.20.1 was released to fix the DLL name.

**Comment by mlocati** [comment]
I just checked why libintl requires libgcc:

```
C:\>dumpbin /DEPENDENTS libintl-8.dll

Dump of file libintl-8.dll

File Type: DLL

  Image has the following dependencies:

    libiconv-2.dll
    ADVAPI32.dll
    KERNEL32.dll
    msvcrt.dll
    libgcc_s_sjlj-1.dll

C:\>dumpbin /IMPORTS:libgcc_s_sjlj-1.dll libintl-8.dll

Dump of file libintl-8.dll

File Type: DLL

  Section contains the following imports:

    libgcc_s_sjlj-1.dll
              61EDB37C Import Address Table
              61EDB1F4 Import Name Table
                     0 time date stamp
                     0 Index of first forwarder reference

                   78 __udivdi3
                   7A __umoddi3
```

So it seems that libgcc is required only because libintl requires `__udivdi3` and `__umoddi3`: maybe we could find a replacement of these functions so that we no more need libgcc (we won't need the ugly `LIBS` patch, and the size of libintl will be reduced by ~500 KB).

**Comment by k-takata** [comment]
I don't think that those functions are used in the source code explicitly. They might be used implicitly by the compiler.

**Comment by k-takata** [comment]
It seems that the functions are used for 64-bit division and modulo internally.

**Comment by mlocati** [comment]
I'll try to ask in the gettext newsletter... I'm a bit reluctant to add the patch:

- this LIBS patch seems really a bad hack to me (yes, I know it's the only solution we have so far)
- other binaries will still require libgcc (libgettextlib, libgettextpo, libgettextsrc, libtextstyle, msggrep), so I'd still need to keep it in the distribution files

**Comment by mlocati** [comment]
PS: I've checked that libgettextlib, libgettextpo, libgettextsrc, libtextstyle, msggrep only use libgcc for these functions:
- `__divdi3`
- `__udivdi3`
- `__umoddi3`

**Comment by mlocati** [comment]
PS: here's the full list of system library dependencies (for shared builds):

# 32 bits

- `libwinpthread-1.dll`
  required by
  - `bin\libgettextlib-0-20-1.dll` (`clock_gettime`)
  - `bin\libstdc++-6.dll`
- `libgcc_s_sjlj-1.dll`
  required by
  - `bin\libgettextlib-0-20-1.dll` (`__divdi3`, `__udivdi3`, `__umoddi3`)
  - `bin\libgettextpo-0.dll` (`__udivdi3`, `__umoddi3`)
  - `bin\libgettextsrc-0-20-1.dll` (`__udivdi3`, `__umoddi3`)
  - `bin\libintl-8.dll` (`__udivdi3`, `__umoddi3`)
  - `bin\libstdc++-6.dll`
  - `bin\libtextstyle-0.dll` (`__divdi3`, `__udivdi3`, `__umoddi3`)
  - `bin\msggrep.exe` (`__udivdi3`, `__umoddi3`)
- `libstdc++-6.dll`
  required by
  - `bin\xgettext.exe` (`__cxa_guard_acquire`, `__cxa_guard_release`)

# 64 bits

- `libwinpthread-1.dll`
  required by
  - `bin\libgettextlib-0-20-1.dll` (`clock_gettime`)
  - `bin\libstdc++-6.dll`
- `libgcc_s_seh-1.dll`
  required by
  - `bin\libstdc++-6.dll`
- `libstdc++-6.dll`
  required by
  - `bin\xgettext.exe` (`__cxa_guard_acquire`, `__cxa_guard_release`)

**Comment by k-takata** [comment]
How about this patch?

```diff
diff --git a/compile-iconv-gettext-windows.sh b/compile-iconv-gettext-windows.sh
index 2fcb4b2..af557a0 100755
--- a/compile-iconv-gettext-windows.sh
+++ b/compile-iconv-gettext-windows.sh
@@ -517,9 +517,9 @@ bldgtxtCompileGettext () {
     echo '### Making gettext'
     pushd "$BLDGTXT_CONFIGURED/gettext-$BLDGTXT_V_GETTEXT" >/dev/null
     if test -f "libtextstyle/Makefile"; then
-        make $BLDGTXT_MAKE_JOBS --directory=libtextstyle --no-keep-going $BLDGTXT_QUIET_MAKE
+        make $BLDGTXT_MAKE_JOBS --directory=libtextstyle --no-keep-going $BLDGTXT_QUIET_MAKE CCLD='$(CC) -XCClinker -static-libgcc' CXXLD='$(CXX) -XCClinker -static-libgcc'
     fi
-    make $BLDGTXT_MAKE_JOBS --directory=gettext-tools --no-keep-going $BLDGTXT_QUIET_MAKE
+    make $BLDGTXT_MAKE_JOBS --directory=gettext-tools --no-keep-going $BLDGTXT_QUIET_MAKE CCLD='$(CC) -XCClinker -static-libgcc' CXXLD='$(CXX) -XCClinker -static-libgcc'
     popd >/dev/null
 }
 
```

**Comment by mlocati** [comment]
I tried that, and I now have that `libgcc_s_sjlj-1.dll` is still required by
- `libgettextlib-0-20-1.dll`
- `libgettextsrc-0-20-1.dll`
- `libstdc++-6.dll` (which in turns is required by `xgettext.exe`)

Compared to https://github.com/mlocati/gettext-iconv-windows/issues/8#issuecomment-491826834 we have less files that depends on it, but I'd still need to include `libgcc`.

I guess that the only solution for you is to fork this project, and save [your patch](https://github.com/mlocati/gettext-iconv-windows/issues/8#issuecomment-491387524) as `patches/gettext-0.20.1-configure/shared/000-no-libgcc-for-libintl.patch`...

**Comment by mlocati** [comment]
Vincent Torri [wrote](http://lists.gnu.org/archive/html/bug-gettext/2019-05/msg00102.html) about the `__USE_MINGW_ANSI_STDIO` define.
If we use `-D__USE_MINGW_ANSI_STDIO=0` in the `configure` command, it seems that libintl does not require anymore libgcc :wink:

**Comment by mlocati** [comment]
I've updated this repository with a few changes.

Here's the list of dependencies for shared builds that we now have:

```
# 32 bits
- libgcc_s_sjlj-1.dll required by
  - bin\libgettextlib-0-20-1.dll
    imported functions: __udivdi3, __umoddi3
  - bin\libgettextpo-0.dll
    imported functions: __udivdi3
  - bin\libstdc++-6.dll
- libstdc++-6.dll required by
  - bin\xgettext.exe
    imported functions: __cxa_guard_acquire, __cxa_guard_release
- libwinpthread-1.dll required by
  - bin\libgcc_s_sjlj-1.dll
  - bin\libstdc++-6.dll

# 64 bit
- libgcc_s_seh-1.dll required by
  - bin\libstdc++-6.dll
- libstdc++-6.dll required by
  - bin\xgettext.exe
    imported functions: __cxa_guard_acquire, __cxa_guard_release
- libwinpthread-1.dll required by
  - bin\libgcc_s_seh-1.dll
  - bin\libstdc++-6.dll
```

So, it seems that libgcc is no more required by libintl.

@k-takata Could you check that everything works for you? You can either compile it (there's a nice [docker way](https://github.com/mlocati/gettext-iconv-windows/blob/master/README.md#building-with-docker) to do it now), or you can check the following attached file.

[gettext0.20.1-iconv1.16-shared-32.zip](https://github.com/mlocati/gettext-iconv-windows/files/3182319/gettext0.20.1-iconv1.16-shared-32.zip)

**Comment by k-takata** [comment]
I could totally remove the dependency on libgcc and libstdc++ from all files with this dirty hack: https://github.com/k-takata/gettext-iconv-windows/commit/a7355c23f2d96a6705133fb7c08d39da446540df

libgettextpo-0.dll's dependency on libgcc was removed by adding `CCLD='$(CC) -XCClinker -static-libgcc'` option to `make`.
libgettextlib-0-20-1.dll's dependency on libgcc was removed by removing the `-lgcc_s` option from the libtool's command line. (Capture the output of libtool and strip `-lgcc_s` by sed, and re-execute the command.)
xgettext.exe's dependency on libstdc++-6.dll was removed by using gcc as a linker (instead of g++) and adding `-XCClinker "-Wl,-Bstatic -lstdc++ -lgcc -Wl,-Bdynamic"` option.

**Comment by k-takata** [comment]
When I tried your latest source, `0001-Avoid-dependency-on-libwinpthread.patch` (added by 0abe54a04a225e0161166640b30c6a0fce09300b) failed to be applied.
`gettext-tools/config.h` doesn't exist before running the `configure` script.

**Comment by mlocati** [comment]
> `gettext-tools/config.h` doesn't exist before running the `configure` script.

Yep, the `gettext-tools/config.h` file is created by the `configure` command, and that's why it's stored in the `patches/gettext-0.20.1-make` folder instead of `patches/gettext-0.20.1-configure`

**Comment by k-takata** [comment]
Ah, I got it. But I rerun the build script, I got the following message:

```
 - patching (0001-Avoid-dependency-on-libwinpthread.patch)
1 out of 1 hunk FAILED -- saving rejects to file gettext-tools/config.h.rej
```

Looking at the `gettext-tools/config.h`, it is as follows:

```c
#define HAVE_CHSIZE 1

/* Define to 1 if you have the `clock_gettime' function. */
/* #undef HAVE_CLOCK_GETTIME */

/* Define to 1 if you have the `clock_settime' function. */
/* #undef HAVE_CLOCK_SETTIME */
```

Strangely, `HAVE_CLOCK_GETTIME` is not defined.

**Comment by mlocati** [comment]
Do you run it with the code of the `master` branch, or with your patches?

**Comment by k-takata** [comment]
On the master branch without my patches.

**Comment by mlocati** [comment]
What about https://github.com/mlocati/gettext-iconv-windows/commit/1894718ec70911b8e693ba76dec6dd56d3081cc5 ?

**Comment by k-takata** [comment]
Thank you. It works.

**Comment by k-takata** [comment]
Slightly updated my patch: https://github.com/k-takata/gettext-iconv-windows/commit/59ecb989b5ede115ed2fe7ca5b400376367bf4e2

**Comment by bhaible** [comment]
As mentioned on the bug-gettext mailing list, I've made two changes to reduce the shared library dependencies:

- https://git.savannah.gnu.org/gitweb/?p=gettext.git;a=commitdiff;h=45500ab1765581d6a3b7d2e6a6c2595466de70af removes the dependency towards libgcc_s-sjlj for libintl, libasprintf, libtextstyle.
- https://git.savannah.gnu.org/gitweb/?p=gettext.git;a=commitdiff;h=bf9fd4eb0a052b9675b8385030f7a87a2743520b removes the dependency towards libwinpthread from libintl at least.

**Comment by mlocati** [comment]
Thanks Bruno! I guess that https://git.savannah.gnu.org/gitweb/?p=gettext.git;a=commitdiff;h=bf9fd4eb0a052b9675b8385030f7a87a2743520b also requires https://git.savannah.gnu.org/gitweb/?p=gettext.git;a=commitdiff;h=a4988721cf6e636755b64ac9d161ec95e7b79e64 isn't it?

**Comment by bhaible** [comment]
Yes.

**Comment by k-takata** [comment]
It seems that this has been fixed in [gettext 0.20.2 + iconv 1.16](https://github.com/mlocati/gettext-iconv-windows/releases/tag/v0.20.2-v1.16). Closing.

**Comment by bhaible** [comment]
Just checked the current situation (with gettext-0.26-pre1, built without any workarounds):

* 32-bit

  libtextstyle needs libgcc_s_sjlj-1.dll for
    __udivdi3
    __udivmoddi4
  
  libgexttextlib needs libgcc_s_sjlj-1.dll for
    __divdi3
    __udivdi3
    __udivmoddi4

  xgettext needs libgcc_s_sjlj-1.dll for
    __udivdi3

  xgettext needs libstdc++-6.dll for
    __cxa_guard_acquire
    __cxa_guard_release
    
* 64-bit

  xgettext needs libstdc++-6.dll for
    __cxa_guard_acquire
    __cxa_guard_release

**Comment by mlocati** [comment]
I've just [tried to build gettext 0.26-pre1](https://github.com/mlocati/gettext-iconv-windows/pull/49), and here's what we have for [shared 32-bit builds](https://github.com/mlocati/gettext-iconv-windows/actions/runs/16143809150/job/45557347174#step:50:1):

```
libgcc_s_sjlj-1.dll added because:
  - libgettextlib-0-26-pre1.dll uses its functions: __divdi3, __udivdi3, __udivmoddi4
  - libintl-8.dll uses its functions: __udivdi3, __udivmoddi4
  - libtextstyle-0.dll uses its functions: __udivdi3, __udivmoddi4
  - xgettext.exe uses its functions: __udivdi3
libwinpthread-1.dll added because:
  - libgcc_s_sjlj-1.dll requires it
```



> * xgettext needs libstdc++-6.dll for
>       __cxa_guard_acquire
>       __cxa_guard_release

My builds don't require `libstdc++-6.dll` because [I use `CXXFLAGS='-fno-threadsafe-statics'`](https://github.com/mlocati/gettext-iconv-windows/blob/v0.25.1-v1.17/build-exe/vars.ps1#L233C23-L233C47) - see:
- https://sourceforge.net/p/mingw-w64/mailman/message/58824383/
- https://lists.gnu.org/archive/html/bug-gettext/2024-10/msg00008.html
- https://lists.gnu.org/archive/html/bug-gettext/2024-10/msg00010.html

**Comment by mlocati** [comment]
@k-takata I've just published gettext 1.0.

Please note that in this version `libintl-8.dll` [ doesn't depend on `libgcc_s_sjlj-1.dll`](https://github.com/mlocati/gettext-iconv-windows/actions/runs/21480028099/job/61873799233#step:53:60) anymore:

```sh
$ objdump -p libintl-8.dll | grep 'DLL Name'
        DLL Name: libiconv-2.dll
        DLL Name: ADVAPI32.dll
        DLL Name: KERNEL32.dll
        DLL Name: msvcrt.dll

$ objdump -p libiconv-2.dll | grep 'DLL Name'
        DLL Name: KERNEL32.dll
        DLL Name: msvcrt.dll
```

**Comment by k-takata** [comment]
Thank you. That's nice.
